File: src/utils/presentation.ts
import { v4 } from '@lukeed/uuid'

import { EJSON } from '../ejson'

export enum PayloadType {
  METHOD = 'method',
  RESULT = 'result',
  EVENT = 'event',
  ERROR = 'error',
  AUTH_RESULT = 'auth:result',
}

export namespace Presentation {
  export const uuid = v4

  export type Payload = {
    type: PayloadType
    [key: string]: any
  }

  export function decode<T = Payload>(payload: string | { data: string }): T {
    return EJSON.parse(
      typeof payload === 'string'
        ? payload
        : (payload as { data: string }).data,
    )
  }

  export function encode<T = Payload>(payload: T): string {
    return EJSON.stringify(payload)
  }
}

File: bifrost-py/src/bifrost/ejson/presentation.py
"""``Presentation.encode`` / ``Presentation.decode`` — port of
``bifrost-ts/src/utils/presentation.ts``.

Every HTTP body, WebSocket frame, and Redis message on the wire goes
through this layer. See PROTOCOL.md §3.
"""

from __future__ import annotations

import uuid as _uuid
from typing import Any

from bifrost.ejson.parse_stringify import parse, stringify


class Presentation:
    """Namespace object mirroring the TS ``Presentation`` export."""

    @staticmethod
    def encode(payload: Any) -> str:
        """Serialize ``payload`` to an EJSON-text string."""

        return stringify(payload)

    @staticmethod
    def decode(payload: str | dict[str, Any]) -> Any:
        """Decode an EJSON-text string, or unwrap ``{"data": "<ejson>"}``."""

        if isinstance(payload, str):
            return parse(payload)
        if isinstance(payload, dict) and "data" in payload:
            return parse(payload["data"])
        raise TypeError("Presentation.decode expects a string or {'data': str}")

    @staticmethod
    def uuid() -> str:
        """Generate a fresh UUIDv4 string (matches TS ``Presentation.uuid``)."""

        return str(_uuid.uuid4())
